tests/test_qir/test_compile.py

```diff
@@ -0,0 +1,538 @@
+# Copyright 2023 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import inspect
+import re
+from operator import attrgetter
+from typing import Callable
+
+import pytest
+
+
+from dwave.gate import Circuit
+from dwave.gate import operations as ops
+
+pyqir = pytest.importorskip("pyqir")
+# QIR submodule imports must postcede PyQIR importorskip
+from dwave.gate.qir.compiler import BaseModule, CompileError, qir_module  # noqa: E402
+from dwave.gate.qir.instructions import InstrType, Instruction, Operations  # noqa: E402
+
+
+@pytest.fixture(scope="function")
+def two_op_module():
+    """Module with an ``X``, a ``CNOT`` and a ``Measurement`` operation."""
+    mod = BaseModule("QIR_module", 2, 1)
+
+    args_op_0 = (mod.qubits[0],)
+    args_op_1 = (mod.qubits[0], mod.qubits[1])
+    args_meas = (mod.qubits[0], mod.results[0])
+
+    instruction_0 = Instruction(InstrType.base, "x", args_op_0)
+    instruction_1 = Instruction(InstrType.base, "cx", args_op_1)
+    instruction_2 = Instruction(InstrType.measurement, "mz", args_meas)
+
+    mod.add_instruction("body", instruction_0)
+    mod.add_instruction("body", instruction_1)
+    mod.add_instruction("measurements", instruction_2)
+
+    return mod
+
+
+def assert_equal(str_0: str, str_1: str) -> bool:
+    """Assert that two multi-line strings are equal."""
+    for r, e in zip(str_0.split("\n"), str_1.split("\n")):
+        # strip whitespace from each line and check
+        assert r.strip() == e.strip()
```

Comment:
Any reason to prefer that instead? I actually find just `assert ... == ...` clearer.

Reply:
Sorry if that was unclear. What I meant:

> To see how your `assert_equal()` is broken, consider a case of inputs with unequal lengths, for example: `assert_equal('a', 'a\nb')` -- where `str_0` has one line and `str_1` has two lines. Because `zip` stops on _any_ iterator exhausted, the second line will be silently ignored, and only the first lines will be compared.

Reply:
I see, sorry, I somehow missed that the function was called `assert_equal`. 🤦